transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java

```diff
@@ -817,8 +803,24 @@ public ChannelFuture writeAndFlush(Object msg) {
         return writeAndFlush(msg, newPromise());
     }
 
-    private static void notifyOutboundHandlerException(Throwable cause, ChannelPromise promise) {
-        PromiseNotificationUtil.tryFailure(promise, cause, logger);
+    private void handleOutboundHandlerException(Throwable cause, boolean closeDidThrow, ChannelPromise promise) {
+        String msg = handler() + " threw an exception while handling an outbound event. This is most likely a bug";
+
+        if (promise != null && !promise.isDone()) {
+            // This is a "best-effort" approach to at least ensure the promise will be completed somehow.
+            promise.tryFailure(new IllegalStateException(msg, cause));
+        }
+
+        logger.warn("{}. This is most likely a bug, closing the channel.", msg, cause);
+        if (closeDidThrow) {
```

Comment:
The `close()` calls below can fail too, should we just add a defensive `try-catch`, log and return?

Reply:
Not sure how it could fail ... if it fails it should be in the next context

Reply:
I see what you are saying; if we assume that nothing apart from the handler can throw then you are right. Not sure if it is worth considering custom context wrappers, etc.

Reply:
Yep I think it should be fine without a guard